### ISSUE TITLE
Support of registration marks

### ISSUE THREAD

**Issue by stephanedaviet** [fablabnbg/inkscape-silhouette#20]
It's more a question than an issue: are registration marks supported? I see some mention of them in the code, but there's no clear evidence of this. Moreover, it is not stated neither it is supported nor it isn't in the Readme.

**Comment by skruppy** [comment]
I'm also interested in this topic. I found nosliwneb/robocut, which has support for registration marks [implemented](https://github.com/nosliwneb/robocut/blob/7cef69287c5b59ca4d4228fac2477aebbc7fbbc7/Plotter.cpp#L445).

**Comment by probonopd** [comment]
:+1:

**Comment by jnweiger** [comment]
There is some code for registration marks in the python implementation, but I never got it working.
I don't understand how the registration mark mode is supposed to work, unfortunately.

Great to know that robocut has an implementation. Can you describe how it is used with robocut?
I plan to port the code to Java for the next iteration....

**Comment by skruppy** [comment]
This is an untested!! adaption of my previously linked rebocut reference: https://github.com/Skrupellos/inkscape-silhouette/tree/regmarks

> Great to know that robocut has an implementation. Can you describe how it is used with robocut?

Like this:

![AltText](http://s16.postimg.org/v0gzgsyxh/Untitled.png)

(and you have to add the registration marks on your own. There are [Templates](https://github.com/Timmmm/robocut/tree/master/examples) but they seem to be incompatible with my cutter)

> I plan to port the code to Java for the next iteration....

OK, the code is ugly, but such a drastic measure xD

**Comment by jnweiger** [comment]
@stephanedaviet Please checkout the 1.17 release or later, maybe your wish has come true there. https://github.com/fablabnbg/inkscape-silhouette/releases/tag/v1.17

**Comment by probonopd** [comment]
Very cool. Is it working for you and which Template did you use?

**Comment by jnweiger** [comment]
I don't know how to use regmarks, sorry. Did not even try with the original software. We need to wait for other testers :-)

**Comment by jnweiger** [comment]
Code added. Please reopen if testing exposes issues.

**Comment by probonopd** [comment]
I can test on a silhouette portrait if you tell me how this is supposed to work ;-)

**Comment by stephane-daviet-maif** [comment]
I will take time to test it and to explain how it works in the next week.

**Comment by EtherGraf** [comment]
I have tested the recent version 1.17 with my silhouette portrait. It makes some movements and at the expected registration marks, seems to search on the correct place. Btw. I have to swap the width and height values for the regmark. 
But the result is "Regmarks not found".

Also I'm not sure of difference of "Document has regmarks" and "Search for regmarks". Is this not the same meaning?

**Comment by EtherGraf** [comment]
I have just done some experiments with the protocol and found some facts. 

Initialising: 
FN0 -- Sets Coordinate system, 0 =LeftToRight, 1=RightToLeft
TB50,0 -- Reset?
TB99 -- Job done?

---

Registration marks:
TB52,2 -- Type of Marks: 0=Original,SD / 2=Cameo,Portrait
TB51,400 -- Length of Registration Mark = 20mm
TB53,20 -- Width of Registration Mark = 1mm
TB55,1 -- Always One
TB23,5140,3400 Search Registration Marks manual, 5140=height, 3400=width
TB123,5140,3400,200,158 Search Registration marks automatic, 5140=height, 3400=width, 200=top(0=10mm, 200=10mm), 158=left (58=15mm, -42=10mm, there is always an additional border of 12,1mm!)
TB99

Today I have no more time to comment this in sourcecode, maybe next days.

**Comment by EtherGraf** [comment]
Repeated status response while searching for regmarks is:

"    1" or "1" Busy
"    0" or "0"  Ready (successfull found)
"    -1" Error (not found)
"2" No paper loaded

Note that leading blanks can occur!

**Comment by EtherGraf** [comment]
Just for your information:
I have sucessfully plotted with autoscanned registration marks on a Silhouette Portrait with Ubuntu 14.04.
Find the tested code on the branch 'registrationmarks'.

**Comment by stephane-daviet-maif** [comment]
I have tested versions of code both from @Skrupellos (PR #23) and @EtherGraf (from the _registrationmarks_ branch of his fork https://github.com/EtherGraf/inkscape-silhouette).

PR #23 doesn't work for me or I do not have the right template file for registration marks. Like @EtherGraf, it always end with “"Regmarks not found”.

Code for @EtherGraf seems to work better. Registration marks are found with the template he provides in his repo (https://github.com/EtherGraf/inkscape-silhouette/blob/registrationmarks/examples/registration-marks.svg). But, then, my cuttings are biased. For instance, while trying to cut rectangles aligned with the borders of my sheet, the Silhouette cuts instead some parallelograms: borders which are parallel with long edge are ok, borders which should be parallel to short edge are cut diagonally.

I'm seeking in the code to see if I can spot the origin of this behaviour.

**Comment by EtherGraf** [comment]
@stephanedaviet: Are you working with a silhouette portrait or a silhouette cameo device?
I'm using a portrait and it has worked so far. I think, for the cameo some other registration marks are needed.

A simple shifting bias could be compensated with the offset values. In general it has to be zero  while working with registration marks, because the compensation should be done on the device.
For parallelogram errors I have no reasonable explanation. How big is the error?

**Comment by EtherGraf** [comment]
@stephanedaviet  In my released template the positions of the regmarks was slightly off. With this I got an error similarly to your description. It was an parallelogram with 0,3mm offset in both directions.
I have checked in an corrected version to github in the regmarks branch.

While testing precise cuttings I found out, that on my device there seems to be an offset between the regmark searching optics and the cutting knife, so I have to set an additonal offset for both x and y to 0,1mm on the first tab. Don't expect to much precision from these devices!
